tkg/clusterclient/clusterclient.go

```diff
@@ -2505,44 +2483,32 @@ func (c *client) UpdateAWSCNIIngressRules(clusterName, clusterNamespace string)
 	}
 
 	cniIngressRules := awsCluster.Spec.NetworkSpec.CNI.CNIIngressRules
-	cniIngressRulesNeedUpdate := false
-	for _, ingressRule := range AWSIngressRules {
-		if !containsIngressRule(cniIngressRules, ingressRule) {
-			cniIngressRules = append(cniIngressRules, ingressRule)
-			cniIngressRulesNeedUpdate = true
-		}
-	}
-
-	if cniIngressRulesNeedUpdate {
-		awsCluster.Spec.NetworkSpec.CNI.CNIIngressRules = cniIngressRules
-		err := retry.RetryOnConflict(retry.DefaultRetry, func() error {
-			return c.UpdateResource(awsCluster, clusterName, clusterNamespace)
-		})
-		if err != nil {
-			return err
-		}
-	}
-
-	return nil
-}
-
-func containsIngressRule(listOfRules capav1beta2.CNIIngressRules, rule capav1beta2.CNIIngressRule) bool {
-	for _, ingressRule := range listOfRules {
-		if ingressRule.Description != rule.Description {
-			continue
-		}
-		if ingressRule.Protocol != rule.Protocol {
+	// first check if existing ingress rules already contain the new rule
+	for _, ingressRule := range cniIngressRules {
+		if ingressRule.Description != newRule.Description {
 			continue
 		}
-		if ingressRule.FromPort != rule.FromPort {
+
+		if ingressRule.Protocol != newRule.Protocol {
 			continue
 		}
-		if ingressRule.ToPort != rule.ToPort {
+
+		if ingressRule.FromPort != newRule.FromPort || ingressRule.ToPort != newRule.ToPort {
 			continue
 		}
-		return true
+		// If we get here it means the rule exists, so we do nothing.
+		return nil
 	}
-	return false
+
+	// rule was not found so we add it
+	cniIngressRules = append(cniIngressRules, newRule)
+
+	awsCluster.Spec.NetworkSpec.CNI.CNIIngressRules = cniIngressRules
+	err := retry.RetryOnConflict(retry.DefaultRetry, func() error {
+		return c.UpdateResource(awsCluster, clusterName, clusterNamespace)
```

Comment:
Can we use patch instead of update?

Reply:
looking into it.

Reply:
using patch here gets very complex because it is meant to be able to use the function to add rules in the future, which means to use patch we would have to be able to predict what the actual cni rule structure will look like.  I could not come up with a simple way to create the json required for the patch call. But if anyone can think of it I can add it. For now lets stay with UpdateResource, since it had been working in the past.